plasmapy/formulary/distribution.py

```diff
@@ -640,13 +547,17 @@ def Maxwellian_speed_2D(v, T, particle="e", v_drift=0, vTh=np.nan, units="units"
     coeff2 = 2 * np.pi * (v - v_drift)
     expTerm = np.exp(-vSq / vThSq)
     distFunc = coeff1 * coeff2 * expTerm
-    if units == "units":
-        return distFunc.to(u.s / u.m)
-    elif units == "unitless":
-        return distFunc
+    return distFunc
 
 
-def Maxwellian_speed_3D(v, T, particle="e", v_drift=0, vTh=np.nan, units="units"):
+@validate_quantities(T={'equivalencies': u.temperature_energy()})
+def Maxwellian_speed_3D(
+    v: _velocity_unit,
+    T: u.K,
+    particle="e",
+    v_drift: _velocity_unit = 0. * _velocity_unit,
+    vTh: _velocity_unit = np.nan * _velocity_unit,
+) -> (u.s / u.m):
```

Comment:
I think we need to reconcile the use of `T`, `particle`, and `vTh` as arguments here, and all the distribution functions.  First off, the default value for `vTh` should be `None` (i.e. `vTh: _velocity_unit = None`) and then add `none_shall_pass=True` to the decorator.  This is a bit ugly though, since the function still requires you to pass-in `T` and `particle`. ¯\\_(ツ)_/¯

Currently I can only see two possible solutions to this...

1. Make `particle`, `T`, and `vTh` keywords and come up with an elaborate checking mechanism inside the function.  *Still feels ugly to me.*
2. Drop `particle` and `T`, and only require `vTh`.  Then in the docstrings have an explicit example using `parameters.thermal_speed()` to calculate `vTh` .

Reply:
Option 2 definitely sounds cleaner to me, but how many more clients do you think this will break? Your thoughts @StanczakDominik ?

Reply:
Perhaps there's a 3rd option we're missing here. @FilipBolt, in the issue you raised this possible design (I removed the `units` argument):
```python
dist = DistributionFactory.createDistribution(type='maxwellian', dimension=1, property='speed')
dist(v, T, particle="e", v_drift=0, vTh=np.nan)
```
But suppose we did it like this:

```python
dist = DistributionFactory.createDistribution(type='maxwellian', dimension=1, property='speed')
dist.from_temperature(v, T=T, particle="e", v_drift=0)
dist.from_thermal_velocity(v, vTh=vTh, v_drift=0)
```

what do you guys think about this one?

Reply:
Another idea, queuing off of units...

```python
@validate_quantities(
    v={"pass_equivalent_units": True},
    sigma={"pass_equivalent_units": True},
    v_drift={"pass_equivalent_units": True},
)
def Maxwellian_1D(
    v: _velocity_units,
    sigma: [u.K, _velocity_units],
    particle="e",
    v_drift: _velocity_units = 0. * _velocity_units,
):

    if sigma.unit.is_equivalend(_velocity_units):
        # we got a velocity, assume sigma is vTh
        vTh = sigma
    else:
        # we got temperature
        vTh= thermal_speed(sigma, particle=particle, method="most_probable")

    # do calculations

    return distribution
```

I think this is my favorite so far.  Obviously, `sigma` would have to be explicitly explained in the docstrings.

Reply:
@StanczakDominik That looks great. That way you could have different interfaces to specify the distribution from and it is more explicit. 

@rocco8773 I think your solution is fine, but I prefer the one by Dominik, since it is more explicit and requires less documentation reading IMHO. 

Thanks very much for the suggestions. I would like to do this change as part of a subsequent  CR, which will involve refactoring the entire distribution.py to behave like a builder design pattern.

Reply:
Sounds good!

Reply:
The thing is @StanczakDominik's solution does not solve the functional use case.  Yes, it makes sense if working from the factory class, but not working directly with `Maxwellian_1D`.  Directly using the function still requires the user to pass a `T` argument when he or she is trying to use the `vTh` argument.  This will even be a problem for designing the factory class.

I don't like my option 1 solution because I feel it's poor design to make `T` and `Vth` optional arguments, but require the use of one for the function to work.  By definition, they're optional arguments so the function should work without them.

This is why I come back around to my option 3.  It's explicit in the arguments it requires and should only add a few lines to the documentation.

If you want to be more explicit with the naming of the function, then a 4th (no 5th) option would be...

```python
def Maxwellian_1D(v, vTh):
    # do your stuff
    return distribution

def Maxwellian_1D_from_temperature(v, T, particle='e'):
    vTh = thermal_speed(t, particle=particle, method="most_probable")
    return Maxwellian_1D(v, vTh)
```

Reply:
I'd argue that there's probably not much value in devotedly sticking to a functional approach... Only a C programmer deals in this kind of absolutes! ;)

Reply:
That's find, but as a public facing API it shouldn't have bugs like this.  Give it a leading underscore and make the factory class the public facing API, then I'll turn around and walk away grumbling under my breath. 🤣

Reply:
I was thinking to keep the functional version just for backwards compatibility, otherwise I'm   against the fully functional approach mostly because of a clean code approach. 

This is the idea what existing functions would do:
```
def Maxwellian_speed_1D():
    dist = DistributionFactory.createDistribution(type='maxwellian', dimension=1, property='speed')
    return dist.from_thermal_velocity(v, vTh=vTh, v_drift=0)
```

but the idea is to actually avoid allowing clients to do this in practice, but actually create the distribution via the factory, mostly because of software maintenance reasons. You can see that right now, you have 6 veeery similar docstrings and methods in ```distribution.py``` that can be difficult to review if you're making a simple change that might effect all of them (like changing all the signatures).